src/quickunlock/PinUnlock.cpp

```diff
@@ -0,0 +1,171 @@
+            break;
+        }
+    }
+
+    // Hash the pin and use it as the key for the encryption
+    CryptoHash hash(CryptoHash::Sha256);
+    hash.addData(pin.toLatin1());
+    auto key = hash.result();
+
+    // Generate a random IV
+    auto iv = Random::instance()->randomArray(SymmetricCipher::defaultIvSize(SymmetricCipher::Aes256_GCM));
+
+    // Encrypt the data using AES-256-CBC
+    SymmetricCipher cipher;
+    if (!cipher.init(SymmetricCipher::Aes256_GCM, SymmetricCipher::Encrypt, key, iv)) {
+        m_error = QObject::tr("Failed to init KeePassXC crypto.");
+        return false;
+    }
+    QByteArray encrypted = data;
+    if (!cipher.finish(encrypted)) {
+        m_error = QObject::tr("Failed to encrypt key data.");
+        return false;
+    }
+
+    // Prepend the IV to the encrypted data
+    encrypted.prepend(iv);
+    // Store the encrypted data and pin attempts
+    m_encryptedKeys.insert(dbUuid, qMakePair(1, encrypted));
+
+    return true;
+}
+
+bool PinUnlock::getKey(const QUuid& dbUuid, QByteArray& data)
+{
+    data.clear();
+    if (!hasKey(dbUuid)) {
+        m_error = QObject::tr("Failed to get credentials for quick unlock.");
+        return false;
+    }
+
+    const auto& pairData = m_encryptedKeys.value(dbUuid);
+
+    // Restrict pin attempts per database
+    for (int pinAttempts = pairData.first; pinAttempts <= MAX_PIN_ATTEMPTS; ++pinAttempts) {
+        bool ok = false;
+        auto pin = QInputDialog::getText(
+            nullptr,
+            QObject::tr("Quick Unlock Pin Entry"),
+            QObject::tr("Enter quick unlock pin (%1 of %2 attempts):").arg(pinAttempts).arg(MAX_PIN_ATTEMPTS),
+            QLineEdit::Password,
+            {},
+            &ok);
+
+        if (!ok) {
+            m_error = QObject::tr("Pin entry was canceled.");
+            return false;
+        }
+
+        // Hash the pin and use it as the key for the encryption
+        CryptoHash hash(CryptoHash::Sha256);
+        hash.addData(pin.toLatin1());
+        auto key = hash.result();
+
+        // Read the previously used challenge and encrypted data
+        auto ivSize = SymmetricCipher::defaultIvSize(SymmetricCipher::Aes256_GCM);
+        const auto& keydata = pairData.second;
+        auto challenge = keydata.left(ivSize);
+        auto encrypted = keydata.mid(ivSize);
+
+        // Decrypt the data using the generated key and IV from above
+        SymmetricCipher cipher;
+        if (!cipher.init(SymmetricCipher::Aes256_GCM, SymmetricCipher::Decrypt, key, challenge)) {
+            m_error = QObject::tr("Failed to init KeePassXC crypto.");
+            return false;
+        }
+
+        // Store the decrypted data into the passed parameter
+        data = encrypted;
+        if (cipher.finish(data)) {
+            // Reset the pin attempts
+            m_encryptedKeys.insert(dbUuid, qMakePair(1, keydata));
+            return true;
+        }
```

Comment:
No "wrong PIN" error message? No time delay?

Reply:
Not sure I understand this, we re-show the pin dialog with the attempts incremented. If all attempts are exhausted then the error is shown saying so. 

Are you also asking about adding a time-delay attempt reset?